pkg/cmd/cli/sa/newtoken.go

```diff
@@ -0,0 +1,180 @@
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/serviceaccounts"
+	osautil "github.com/openshift/origin/pkg/serviceaccounts/util"
+)
+
+const (
+	NewServiceAccountTokenRecommendedName = "new-token"
+
+	newServiceAccountTokenShort = `Generate a new service account API token for a service account.`
+
+	newServiceAccountTokenLong = `
+Generate a new service account API token for a service account.
+
+Service account API tokens are used by service accounts for privileged access to the API. This command will generate a new
+token, which could be used to compartmentalize service account actions by executing them with distinct tokens, to rotate out
+pre-existing token on the service account, or for use by an external client.
+`
+
+	newServiceAccountTokenUsage = `%s SA-NAME`
+
+	newServiceAccountTokenExamples = `  # Generate a new token and for service account 'custom-sa'
+  $ %[1]s 'custom-sa'
+`
+)
+
+type NewServiceAccountTokenOptions struct {
+	SAName        string
+	Namespace     string
+	SAClient      unversioned.ServiceAccountsInterface
+	SecretsClient unversioned.SecretsInterface
+
+	Out io.Writer
+	Err io.Writer
+}
+
+func NewCommandNewServiceAccountToken(name, fullname string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &NewServiceAccountTokenOptions{
+		Out: out,
+		Err: os.Stderr,
+	}
+
+	newServiceAccountTokenCommand := &cobra.Command{
+		Use:     fmt.Sprintf(newServiceAccountTokenUsage, name),
+		Short:   newServiceAccountTokenShort,
+		Long:    newServiceAccountTokenLong,
+		Example: fmt.Sprintf(newServiceAccountTokenExamples, fullname),
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := options.Complete(args, f); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+			}
+
+			cmdutil.CheckErr(options.Run())
+		},
+	}
+
+	newServiceAccountTokenCommand.Flags().StringVarP(&options.Namespace, "namespace", "n", "", "the namespace that contains the service account")
+	return newServiceAccountTokenCommand
+}
+
+func (o *NewServiceAccountTokenOptions) Complete(args []string, f *clientcmd.Factory) error {
+	if len(args) != 1 {
+		return fmt.Errorf("expected one service account name as an argument, got %q", args)
+	}
+
+	o.SAName = args[0]
+
+	client, err := f.Client()
+	if err != nil {
+		return err
+	}
+
+	if len(o.Namespace) == 0 {
+		namespace, _, err := f.DefaultNamespace()
+		if err != nil {
+			return fmt.Errorf("could not retrieve default namespace: %v", err)
+		}
+		o.Namespace = namespace
+	}
+
+	o.SAClient = client.ServiceAccounts(o.Namespace)
+	o.SecretsClient = client.Secrets(o.Namespace)
+	return nil
+}
+
+// Run creates a new token secret, waits for the service account token controller to fulfill it, then adds the token to the service account
+func (o *NewServiceAccountTokenOptions) Run() error {
+	// declare timeout here, as it is not permitted to take the address of a constant
+	var tokenWatchTimeoutSeconds int64 = 100
```

Comment:
no global constant... not sure a single value would make sense for all watchers. should be an option for the command, e.g. `--timeout`, defaulting either to something sane or to no timeout